### PR TITLE
Optimise spectator loops

### PR DESCRIPTION
**Pull request by SilentSpike**
`count` is only quicker than `forEach` when the final command returns a
boolean or nil. `forEach` is quicker if the final command returns any
other data type.

### CODE REVIEW

File: addons/spectator/functions/fnc_ui_draw3D.sqf [hunk 1]
```diff
@@ -138,9 +132,7 @@ if !(GVAR(uiMapVisible)) then {
                 // Store grenade for next frame
                 _grenadesNew pushBack _x;
             };
-
-            nil // Speed loop
-        } count GVAR(grenadesToDraw);
+        } forEach GVAR(grenadesToDraw);
```
Comment (dedmen) [inline addons/spectator/functions/fnc_ui_draw3D.sqf]: pushBack returns nil so you can still use count

File: addons/spectator/functions/fnc_ui_updateIconsToDraw.sqf [hunk 2]
```diff
@@ -139,9 +139,7 @@ private _camPos = AGLToASL positionCameraToWorld [0,0,0];
             SETVAR(_vehicle,GVAR(firedEH),_vehicle addEventHandler [ARR_2("Fired",{_this call FUNC(handleFired)})]);
         };
     };
-
-    nil // Speed loop
-} count ([] call FUNC(getTargetEntities));
+} forEach ([] call FUNC(getTargetEntities));
```
Comment (dedmen) [inline addons/spectator/functions/fnc_ui_updateIconsToDraw.sqf]: I'm not sure what `if` returns when the then didn't run and there is no `else`. But I think there is a nil on the stack here meaning you can use `count`

File: addons/spectator/functions/fnc_ui_draw3D.sqf (hunk 1 above)
Comment (dedmen) [inline addons/spectator/functions/fnc_ui_draw3D.sqf]: If you swap the pushBack and drawIcon3D you will have a nil on the stack and can then use `count`

**Comment by dedmen** [comment]
>`count` is only quicker than `forEach` when the final command returns a
boolean or nil. `forEach` is quicker if the final command returns any
other data type.
Did you write that yourself or did you copy that from me somewhere?

the benefit from count>forEach is only really important on very big loops that don't do much work. Considering all the time that is spent in each iteration using count here might not even be worth it. The time saved relative to the total time spent.

File: addons/spectator/functions/fnc_ui_updateIconsToDraw.sqf (hunk 2 above)
Comment (dedmen) [reply addons/spectator/functions/fnc_ui_updateIconsToDraw.sqf]: Just looked it up. `if` returns typeless nil when the condition didn't match and there was no else.

**Comment by SilentSpike** [comment]
I paraphrased my understanding of what you'd said in slack the other day. Though yeah I might just change them all to `forEach` as I imagine the performance difference is going to be negligible (until we start using intercept for stuff).

**Comment by SilentSpike** [comment]
Just made them all `forEach` because the performance impact is meaningless. If we eventually start using intercept for stuff, I believe @dedmen has a much faster iteration method we can adopt.

**Pull request merged**